examples/hetero/hetero_unsupervised_dblp.py

```diff
@@ -0,0 +1,187 @@
+import numpy as np
+from torch.nn import Parameter
+from sklearn.model_selection import train_test_split
+from sklearn.linear_model import LogisticRegression
+import os.path as osp
+import math
+import torch
+import torch.nn.functional as F
+from torch_geometric.datasets import DBLP
+from torch_geometric.nn import SAGEConv, HeteroConv, GATConv
+
+EPS = 1e-15
+
+path = osp.join(osp.dirname(osp.realpath(__file__)), '../../data/DBLP')
+dataset = DBLP(path)
+data = dataset[0]
+print(data)
+
+# We initialize conference node features with a single feature.
+data['conference'].x = torch.ones(data['conference'].num_nodes, 1)
+
+
+def uniform(size, tensor):
+    if tensor is not None:
+        bound = 1.0 / math.sqrt(size)
+        tensor.data.uniform_(-bound, bound)
+
+
+class GATEncoder(torch.nn.Module):
+    def __init__(self, in_channels, out_channels, heads):
+        super(GATEncoder, self).__init__()
+        self.conv = GATConv(in_channels, out_channels, heads)
+        self.activation = torch.nn.PReLU()
+
+    def forward(self, x: torch.Tensor, edge_index: torch.Tensor):
+        return self.activation(self.conv(x, edge_index))
+
+
+class HeteroUnsupervised(torch.nn.Module):
+    def __init__(self, metadata, hidden_channels, out_channels, num_layers):
+        super().__init__()
+
+        # for discriminator
+        self.embed_size = out_channels
+        self.weight = Parameter(torch.Tensor(out_channels, out_channels))
+
+        # for encoder
+        self.encoder = GATEncoder(hidden_channels, out_channels, heads=1)
+
+        self.convs = torch.nn.ModuleList()
+        for _ in range(num_layers):
+            conv = HeteroConv({
+                edge_type: SAGEConv((-1, -1), hidden_channels)
+                for edge_type in metadata[1]
+            })
+            self.convs.append(conv)
+
+    def forward(self, xx_dict, edge_index_dict):
+        x_dict = {}
+        for conv in self.convs:
+            x_dict = conv(xx_dict, edge_index_dict)
+            x_dict = {key: F.leaky_relu(x) for key, x in x_dict.items()}
+
+        target_embeds = x_dict['author']
+        # using meta path author paper author
+        edge_index_a = edge_index_dict[('author', 'to', 'paper')]
+        edge_index_b = edge_index_dict[('paper', 'to', 'author')]
+
+        target_edge_index = \
+            self.convert_to_meta_path_edge_index(edge_index_a, edge_index_b)
+
+        pos_feat = target_embeds
+        # shuffle feature corruption
+        idxx = np.random.permutation(pos_feat.shape[0])
+        corrupted_feat = pos_feat[idxx]
+
+        pos_embed = self.encoder(pos_feat, target_edge_index)
+
+        neg_embed = self.encoder(corrupted_feat, target_edge_index)
+
+        summary = torch.mean(pos_embed, dim=0)
+
+        return pos_embed, neg_embed, summary
+
+    def convert_to_meta_path_edge_index(self, edge_index_a, edge_index_b):
+        paper2author_dict = \
+            {idx_a.item(): idx_b.item() for idx_a, idx_b in edge_index_b.T}
```

Comment:
It seems like you are only holding one neighbor, and dropping all the remaining ones. Is this intended?

Reply:
It is intended.
I just added one meta path `author-paper-author(APA)`, we can also add `author-paper-conference-paper-author(APCPA)`, etc. It requires more codes though.
This meta path method is actually from DMGI from AAAI2020, https://www.sciencedirect.com/science/article/abs/pii/S0950705120302240 . 

But it seems a little bit untidy as it requires pretty much preprocessing on data edge index, and I'm doing a simpler method paper method, but as mentioned before it's not finished yet.

Reply:
In case this is intended, you can create a `[num_authors, 2]` matrix and simply scatter paper indices pointing to a particular source/destination author nodes, e.g., something like:
```python
out = torch.empty((num_authors, 2), dtype=torch.long)

out[0].scatter_(edge_index_a[0], edge_index_a[1])
out[1].scatter_(edge_index_b[1], edge_index_b[0])
```